lib/src/dbs/capabilities.rs

```diff
@@ -0,0 +1,479 @@
+pub enum NetTarget {
+	Host(url::Host<String>, Option<u16>),
+	IPNet(ipnet::IpNet),
+}
+
+impl Target for NetTarget {
+	fn matches(&self, elem: &Self) -> bool {
+		match self {
+			// If self contains a host and port, the elem must match both the host and port
+			Self::Host(host, Some(port)) => match elem {
+				Self::Host(_host, Some(_port)) => host == _host && port == _port,
+				_ => false,
+			},
+			// If self contains a host but no port, the elem must match the host only
+			Self::Host(host, None) => match elem {
+				Self::Host(_host, _) => host == _host,
+				_ => false,
+			},
+			// If self is an IPNet, it can match both an IPNet or a Host elem that contains an IPAddr
+			Self::IPNet(ipnet) => match elem {
+				Self::IPNet(_ipnet) => ipnet.contains(_ipnet),
+				Self::Host(host, _) => match host {
+					url::Host::Ipv4(ip) => ipnet.contains(&IpAddr::from(ip.to_owned())),
+					url::Host::Ipv6(ip) => ipnet.contains(&IpAddr::from(ip.to_owned())),
+					_ => false,
+				},
+			},
+		}
+	}
+}
+
+impl std::str::FromStr for NetTarget {
+	type Err = String;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		// If it's a valid IPNet, return it
+		if let Ok(ipnet) = s.parse::<IpNet>() {
+			return Ok(NetTarget::IPNet(ipnet));
+		}
+
+		// If it's a valid IPAddr, return it as an IPNet
+		if let Ok(ipnet) = s.parse::<IpAddr>() {
+			return Ok(NetTarget::IPNet(IpNet::from(ipnet)));
+		}
+
+		// Parse the host and port parts from a string in the form of 'host' or 'host:port'
+		if let Ok(url) = Url::parse(format!("http://{s}").as_str()) {
+			if let Some(host) = url.host() {
+				// Url::parse will return port=None if the provided port was 80 (given we are using the http scheme). Get the original port from the string.
+				if let Some(Ok(port)) = s.split(':').last().map(|p| p.parse::<u16>()) {
+					return Ok(NetTarget::Host(host.to_owned(), Some(port)));
+				} else {
+					return Ok(NetTarget::Host(host.to_owned(), None));
+				}
+			}
+		}
+
+		Err(format!(
+			"The provided network target `{s}` is not a valid host, ip address or ip network"
+		))
+	}
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Targets<T: Target + Hash + Eq + PartialEq> {
+	None,
+	Some(HashSet<T>),
+	All,
+}
+
+impl<T: Target + Hash + Eq + PartialEq + std::fmt::Debug> Targets<T> {
+	fn matches(&self, elem: &T) -> bool {
+		match self {
+			Self::None => false,
+			Self::All => true,
+			Self::Some(targets) => targets.iter().any(|t| t.matches(elem)),
+		}
+	}
+}
+
+/// Capabilities are used to limit what a user can do to the system.
+///
+/// Capabilities are split into 3 categories:
+/// - Scripting: Whether or not the user can execute scripts
+/// - Functions: Whether or not the user can execute certain functions
+/// - Network: Whether or not the user can access certain network addresses
+///
+/// Capabilities can be configured globally or per context based on the Namespace/Database/Scope configuration. The default capabilities are:
```

Comment:
How would `per context based on the Namespace/Database/Scope configuration` be done? Or is that outside of the scope of this PR?

Reply:
Yeah, outside of the scope of this PR.

```suggestion
/// Capabilities are configured globally. By default, capabilities are configured as:
```